### PR TITLE
feat: add envforge audit subcommand (#181)

### PR DESCRIPTION
**Pull request by Yuthika10**
## Description

First PR of a three-part series for #181. Splitting the audit feature into three reviewable PRs rather than one large one:

1. **This PR — MVP foundation:** CLI subcommand, `LocalEnvironment` + `LockfileSource`, text output, semver severity, tests
2. **Next PR:** `--json` and `--sarif` output formats
3. **PR after that:** `ConfigFileSource`, `RemoteApiSource` (once #85 lands), drift score, and compatibility-engine-based severity

If you'd rather see different cuts (e.g. all output formats in one PR, sources in another), happy to restructure.

---

`envforge audit <source-a> <source-b>` prints a Rich-formatted table of differences classified by severity (major / minor / patch / added / removed).

```bash
# Compare local install against a lockfile
envforge audit local requirements.txt

# Compare two lockfiles
envforge audit requirements.lock requirements.txt
```

For an identical pair, output is `No drift detected (N packages match)`. For drift:

```
Audit: lockfile:a.txt -> lockfile:b.txt

╭──────────┬─────────────────┬─────────────────┬──────────╮
│ Package  │ lockfile:a.txt  │ lockfile:b.txt  │ Severity │
├──────────┼─────────────────┼─────────────────┼──────────┤
│ django   │ 4.2.0           │ 5.0.0           │ major    │
│ requests │ 2.31.0          │ 2.32.0          │ minor    │
│ numpy    │ 1.24.0          │ 1.24.1          │ patch    │
╰──────────┴─────────────────┴─────────────────┴──────────╯

Summary: 3 differences (1 major, 1 minor, 1 patch); 0 matching packages.
```

### Source types in this PR

| Spec | Source class | Description |
|---|---|---|
| `local` | `LocalEnvironment` | Active Python interpreter via `pip list --format=json` |
| path to file | `LockfileSource` | requirements.txt-format pinned lines |

`LockfileSource` handles realistic content: inline `#` comments, blank lines, flag lines (`-r requirements-dev.txt`, `--index-url`), package extras (`pkg[extra]==1.0`), and environment markers (`pkg==1.0; python_version<'3.10'`). Unpinned lines are skipped since they can't be meaningfully diffed.

### Scope: this PR vs follow-ups

The issue's full scope splits cleanly into multiple PRs. This one is the foundation.

**In this PR:** CLI subcommand, two source types, text output, semver severity, 22 unit tests.

**Intentional follow-ups (each its own PR):**
- `ConfigFileSource` and `RemoteApiSource` (the latter depends on #85)
- JSON output via `--json`
- SARIF output via `--sarif`
- Drift score + compatibility-engine-based severity (currently semver heuristic)
- `--strict` flag for non-zero exit on drift (CI use cases)

Happy to file the follow-up issues separately if preferred.

## Related Issues

Implements #181 (first MVP PR; the issue stays open for the remaining scope above).

## Changes Made

- New subpackage `cli/envforge_agent/audit/` (6 files):
  - `models.py` — `Package` (PEP 503 normalized), `DiffEntry`, `AuditResult`
  - `sources.py` — `Source` base + `LocalEnvironment` + `LockfileSource`
  - `differ.py` — `diff()` function + semver severity classification
  - `formatters.py` — Rich text output with severity-styled table
  - `command.py` — Click subcommand
  - `__init__.py` — public exports
- One-line registration in `cli/envforge_agent/cli.py` via `cli.add_command(audit_command)`
- New `cli/tests/test_audit.py` — 22 unit tests
- No new runtime dependencies (uses Click + Rich + stdlib, already in `cli/pyproject.toml`)

## Verification

- [x] Added unit tests (22 tests in `cli/tests/test_audit.py`)
- [x] Ran `pytest tests/` successfully (all 22 pass)
- [x] Manually tested via CLI — `envforge --help` lists the new `audit` subcommand, and end-to-end command behavior is covered by `TestAuditCommand` via Click's `CliRunner`
- [ ] (If applicable) Generated scripts pass `SafetyFilter` — N/A, `audit` is read-only and doesn't generate scripts

Test coverage:
- Package name normalization (PEP 503: case + separator collapsing)
- Version severity classification (major / minor / patch / non-numeric fallback / short versions)
- Lockfile parsing edge cases (comments, blank lines, flag lines, unpinned, extras, environment markers, missing file)
- Diff correctness (identical, added, removed, version-changed)
- End-to-end command invocation via `CliRunner` (success, identical, error cases)

## Documentation

- [ ] Updated `docs/FEATURES.md` — happy to add a section in this PR or a follow-up; let me know your preference
- [ ] Updated `CHANGELOG.md` — same as above
- [x] Code is fully documented and type-hinted (module docstrings + function docstrings + type hints throughout)

Submitted under GSSoC'26.

## Summary by CodeRabbit

* **New Features**
  * Added an audit subcommand to compare environment drift between two sources (active environment or lockfiles).
  * Detects and categorizes added, removed, and version-changed packages with severity (major/minor/patch).
  * Shows formatted, colorized output with per-package differences, severity breakdown, and count of matching packages.

* **Tests**
  * Added extensive tests covering parsing, classification, CLI behavior, and local-environment error handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/rishabh0510rishabh/EnvForage/pull/206?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by rishabh0510rishabh** [comment]
hey @Yuthika10 this is the best pr i have seen till now on this repo, vary clean and explained very well thanks for you contribution. as the PR suggest you can work on this issue in 3 parts no problem from my side you can continue implemanting 2nd part.
before that please consider coderabit review in this pr itself

File: cli/envforge_agent/audit/command.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        sys.exit(2)
+    except FileNotFoundError as exc:
+        err_console.print(f"[ERROR] {exc}")
+        sys.exit(2)
+
+    try:
+        result = diff(src_a, src_b)
+    except (RuntimeError, FileNotFoundError) as exc:
+        err_console.print(f"[ERROR] {exc}")
+        sys.exit(1)
```
[the remark this answers is not in the document]
Comment (Yuthika10) [reply cli/envforge_agent/audit/command.py:67]: Already addressed in the latest commit. LockfileSource now wraps all read failures (FileNotFoundError, UnicodeDecodeError, PermissionError, IsADirectoryError, OSError) as RuntimeError with clear messages, so the simplified `except RuntimeError` here catches all source-level errors uniformly. Tests added for missing-file and invalid-UTF-8 cases.

**Comment by Yuthika10** [comment]
Hi @rishabh0510rishabh, addressed all the CodeRabbit nits in the latest commit (34add27): summary ordering now uses SEVERITY_ORDER, test assertions are case-insensitive, the invalid-source test asserts the error message, and added a missing-interpreter test for LocalEnvironment. All 28 tests pass locally.

Planning to start Part 2 (--json and --sarif output formats) on a fresh branch once you've had a chance to review. There is no rush - I'm happy to wait if you'd prefer to merge this first.

**Review by rishabh0510rishabh — APPROVED** [review]
